src/Microsoft.ApplicationInsights.AspNetCore/DiagnosticListeners/Implementation/HostingDiagnosticListener.cs

```diff
@@ -34,31 +40,67 @@ public HostingDiagnosticListener(TelemetryClient client, ICorrelationIdLookupHel
         public string ListenerName { get; } = "Microsoft.AspNetCore";
 
         /// <summary>
+        /// Determine whether the running AspNetCore Hosting version is 2.0 or higher. This will affect what DiagnosticSource events we receive.
+        /// To support AspNetCore 1.0 and 2.0, we listen to both old and new events.
+        /// If the running AspNetCore version is 2.0, both old and new events will be sent. In this case, we will ignore the old events.
+        /// </summary>
+        public bool IsAspNetCore20
+        {
+            get
+            {
+                if (isAspNetCore20.HasValue)
+                {
+                    return isAspNetCore20.Value;
+                }
+
+                var hostingVersion = typeof(WebHostBuilder).GetTypeInfo().Assembly.GetName().Version;
+                isAspNetCore20 = hostingVersion.Major >= 2;
+
+                return isAspNetCore20.Value;
+            }
+        }
+
+        /// <summary>
+        /// Diagnostic event handler method for 'Microsoft.AspNetCore.Hosting.HttpRequestIn' event.
+        /// </summary>
+        [DiagnosticName("Microsoft.AspNetCore.Hosting.HttpRequestIn")]
+        public void OnHttpRequestIn()
+        {
+            // do nothing, just enable the diagnotic source
+        }
+
+        /// <summary>
+        /// Diagnostic event handler method for 'Microsoft.AspNetCore.Hosting.HttpRequestIn.Start' event.
+        /// </summary>
+        [DiagnosticName("Microsoft.AspNetCore.Hosting.HttpRequestIn.Start")]
+        public void OnHttpRequestInStart(HttpContext httpContext)
+        {
+            if (this.client.IsEnabled())
+            {
+                var requestTelemetry = InitializeRequestTelemetryOnHttpRequestInStart(httpContext);
```

Comment:
I feel this part is a bit long. And I want to give it a function name so people know what it is doing.